app/main.py

```diff
@@ -1,2 +1,132 @@
+from typing import Any, Hashable
+from fractions import Fraction
+from app.node import Node
+
+
 class Dictionary:
-    pass
+    def __init__(self) -> None:
+        self._length: int = 0
+        self._hash_table: list = [None] * 8
+        self._hash_size: int = 8
+        self._resize_factor: Fraction = Fraction(2, 3)
+        self._resize_multiplier: int = 2
+
+    def __setitem__(self, key: Hashable, value: Any) -> None:
+        self._hash_size_check_increase()
+        key_exists, key_hash_position = self._while_key_hash_occupied(
+            action="set", key=key, value=value
+        )
+        if not key_exists:
+            node = Node(hash_=hash(key), key=key, value=value)
+            self._hash_table[key_hash_position] = node
+            self._length += 1
+
+    def __getitem__(self, key: Hashable) -> Any | None:
+        key_exists, return_value = self._while_key_hash_occupied(
+            action="get", key=key
+        )
+        if key_exists:
+            return return_value
+        raise KeyError("Key is not found in the Dictionary")
+
+    def __len__(self) -> int:
+        return self._length
+
+    def __repr__(self) -> str:
+        result = [f"{item.key}:{item.value}"
+                  for item in self._hash_table
+                  if item is not None]
+        return "{" + ", \n".join(result) + "}"
+
+    def __delitem__(self, key: Hashable) -> None:
+        """
+        func to delete item by key if key exists, otherwise raise KeyError
+        """
+        key_hash_position = self._find_key_position(key)
+        index_len = 0
+
+        while index_len <= self._hash_size:
```

Comment:
So why not modify  `_while_key_hash_occupied` method and reuse it here to avoid repetition?

Reply:
Because if we use this method with "del", we cannot garantee that we will look for an item(Node) even if entry point ```table[key_hash_position] == None``` it will immediately exit the loop and give wrong return. It can happen due to collision if we have 3 Nodes with starting point at table[1] for example, they will be written in table[1], table[2], table[3] due to collision and imagine we need to delete first Node at table[1], after that we'll have None, table[2], table[3], now we need to delete table[2] or table[3] but their keys will have same address to table as table[1] that is already deleted, they look at the condition in ```_while_key_hash_occupied``` and see that their entry point (table[1]) is None so they don't even check other cells in table to make sure that needed item even exists.
![NVIDIA_Share_MLxlV766aa](https://github.com/mate-academy/py-dictionary/assets/13889031/5658eadc-0cc4-42c6-ad7c-c7074eebaf3f)

and this condition guarantees us that even if we don't guess exact entry point for our table we'll look at every cells in table at least once (in worst case scenario, if we don't have this key in our table or if every key had same entry point in our table and you had collision every time)
![NVIDIA_Share_50azuCeClz](https://github.com/mate-academy/py-dictionary/assets/13889031/772deaa2-948b-4a2e-bd37-6ab2d246125d)